compparallel.go

```diff
@@ -0,0 +1,298 @@
+			return r.Provide(ctx, cid, provide)
+		},
+	)
+}
+
+// FindProvidersAsync will execute all Routers in parallel, iterating results from them in unspecified oredr.
+// If count is set, only that amount of elements will be returned without any specification about from what router is obtained.
+// To gather providers from a set of Routers first, you can use the ExecuteAfter timer to delay some Router execution.
+func (r *ComposableParallel) FindProvidersAsync(ctx context.Context, cid cid.Cid, count int) <-chan peer.AddrInfo {
+	var totalCount int64
+	ch, _ := getChannelOrErrorParallel(
+		ctx,
+		r.routers,
+		func(ctx context.Context, r routing.Routing) (<-chan peer.AddrInfo, error) {
+			return r.FindProvidersAsync(ctx, cid, count), nil
+		},
+		func() bool {
+			return atomic.AddInt64(&totalCount, 1) > int64(count) && count != 0
+		},
+	)
+
+	return ch
+}
+
+// FindPeer will execute all Routers in parallel, getting the first AddrInfo found and cancelling all other Router calls.
+func (r *ComposableParallel) FindPeer(ctx context.Context, id peer.ID) (peer.AddrInfo, error) {
+	return getValueOrErrorParallel(ctx, r.routers,
+		func(ctx context.Context, r routing.Routing) (peer.AddrInfo, error) {
+			return r.FindPeer(ctx, id)
+		},
+		func(ai peer.AddrInfo) bool {
+			return ai.ID == ""
+		})
+}
+
+// PutValue will execute all Routers in parallel. If a Router fails and IgnoreError flag is not set, the whole execution will fail.
+// Some Puts before the failure might be successful, even if we return an error.
+func (r *ComposableParallel) PutValue(ctx context.Context, key string, val []byte, opts ...routing.Option) error {
+	return executeParallel(ctx, r.routers,
+		func(ctx context.Context, r routing.Routing) error {
+			return r.PutValue(ctx, key, val, opts...)
+		},
+	)
+}
+
+// GetValue will execute all Routers in parallel. The first value found will be returned, cancelling all other executions.
+func (r *ComposableParallel) GetValue(ctx context.Context, key string, opts ...routing.Option) ([]byte, error) {
+	return getValueOrErrorParallel(ctx, r.routers,
+		func(ctx context.Context, r routing.Routing) ([]byte, error) {
+			return r.GetValue(ctx, key, opts...)
+		},
+		func(ai []byte) bool {
+			return len(ai) == 0
+		})
+}
+
+func (r *ComposableParallel) SearchValue(ctx context.Context, key string, opts ...routing.Option) (<-chan []byte, error) {
+	return getChannelOrErrorParallel(
+		ctx,
+		r.routers,
+		func(ctx context.Context, r routing.Routing) (<-chan []byte, error) {
+			return r.SearchValue(ctx, key, opts...)
+		},
+		func() bool { return false },
+	)
+}
+
+func (r *ComposableParallel) Bootstrap(ctx context.Context) error {
+	return executeParallel(ctx, r.routers,
+		func(ctx context.Context, r routing.Routing) error {
+			return r.Bootstrap(ctx)
+		})
+}
+
+func getValueOrErrorParallel[T any](
+	ctx context.Context,
+	routers []*ParallelRouter,
+	f func(context.Context, routing.Routing) (T, error),
+	isEmpty func(T) bool,
```

Comment:
this callback doesn't really seem necessary? can't `f` just return (T, bool, error) and reduce the callback complexity?